pkg/querier/queryrange/queryable.go

```diff
@@ -0,0 +1,120 @@
+package queryrange
+
+import (
+	"context"
+
+	"github.com/cortexproject/cortex/pkg/querier/astmapper"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+const (
+	missingEmbeddedQueryMsg = "missing embedded query"
+	nonEmbeddedErrMsg       = "DownstreamQuerier cannot handle a non-embedded query"
+)
+
+// DownstreamQueryable is an implementor of the Queryable interface.
+type DownstreamQueryable struct {
+	Req     Request
+	Handler Handler
+}
+
+// Querier impls Queryable
+func (q *DownstreamQueryable) Querier(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
+	return &DownstreamQuerier{ctx, q.Req, q.Handler}, nil
+}
+
+// DownstreamQuerier is a an implementor of the Querier interface.
+type DownstreamQuerier struct {
+	Ctx     context.Context
+	Req     Request
+	Handler Handler
+}
+
+// Select returns a set of series that matches the given label matchers.
+func (q *DownstreamQuerier) Select(
+	_ *storage.SelectParams,
+	matchers ...*labels.Matcher,
+) (storage.SeriesSet, storage.Warnings, error) {
+	var embeddedQuery string
+	var isEmbedded bool
+	for _, matcher := range matchers {
+		if matcher.Name == labels.MetricName && matcher.Value == astmapper.EmbeddedQueryFlag {
+			isEmbedded = true
+		}
+
+		if matcher.Name == astmapper.QueryLabel {
+			embeddedQuery = matcher.Value
+		}
+	}
+
+	if isEmbedded {
+		if embeddedQuery != "" {
+			return q.handleEmbeddedQuery(embeddedQuery)
+		}
+		return nil, nil, errors.Errorf(missingEmbeddedQueryMsg)
+
+	}
+
+	return nil, nil, errors.Errorf(nonEmbeddedErrMsg)
+}
+
+// handleEmbeddedQuery defers execution of an encoded query to a downstream Handler
+func (q *DownstreamQuerier) handleEmbeddedQuery(encoded string) (storage.SeriesSet, storage.Warnings, error) {
+	queries, err := astmapper.JSONCodec.Decode(encoded)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ctx, cancel := context.WithCancel(q.Ctx)
+	defer cancel()
+
+	// buffer channels to length of queries to prevent leaking memory due to sending to unbuffered channels after cancel/err
+	errCh := make(chan error, len(queries))
+	samplesCh := make(chan []SampleStream, len(queries))
+	// TODO(owen-d): impl unified concurrency controls, not per middleware
+	for _, query := range queries {
+		go func(query string) {
+			resp, err := q.Handler.Do(ctx, q.Req.WithQuery(query))
+			if err != nil {
+				errCh <- err
+				return
+			}
+			streams, err := ResponseToSamples(resp)
+			if err != nil {
+				errCh <- err
+				return
+			}
+			samplesCh <- streams
```

Comment:
You might also need to select on the `context.Done()`.

Reply:
These channels are also buffered here